src/context/ChainApiContext.tsx

```diff
@@ -1,73 +1,60 @@
 import {ApiPromise, WsProvider} from '@polkadot/api';
-import React, {createContext, useContext, useEffect, useMemo, useState} from 'react';
+import React, {createContext, useContext, useEffect, useReducer} from 'react';
 import {useQueryClient} from 'react-query';
 import {createLogger} from 'src/utils';
 import {NetworkContext} from './NetworkContext';
 
-type ApiChainStatusType = 'unknown' | 'connected' | 'disconnected' | 'ready';
-type ChainApiContextValueType = {
-  inProgress: boolean;
-  status: ApiChainStatusType;
-  api?: ApiPromise;
-};
-
-export const ChainApiContext = createContext<ChainApiContextValueType>({
-  api: undefined,
+const initialState: ChainApiContext = {
   status: 'unknown',
   inProgress: false,
-});
+  api: undefined,
+  wsConnectionIndex: 0,
+};
+
+export const ChainApiContext = createContext<ChainApiContext>(initialState);
 
 export const useApi = () => useContext(ChainApiContext);
 
 const logger = createLogger('ChainApiContext');
 
 export function ChainApiContextProvider({children}: {children: React.ReactNode}) {
-  const [inProgress, setInProgress] = useState(true);
-  const [status, setStatus] = useState<ApiChainStatusType>('unknown');
-  const {currentNetwork} = useContext(NetworkContext);
-  const [api, setApi] = useState<ApiPromise>();
-  const [wsConnectionIndex, setWsConnectionIndex] = useState(0);
+  const [state, dispatch] = useReducer(reducer, initialState);
 
+  const {currentNetwork} = useContext(NetworkContext);
   const queryClient = useQueryClient();
 
   useEffect(() => {
-    const wsAddress = currentNetwork.ws[wsConnectionIndex];
+    const wsAddress = currentNetwork.ws[state.wsConnectionIndex];
     if (!wsAddress) {
       return;
     }
 
-    setApi(undefined);
-    setInProgress(true);
     logger.debug('ChainApiContext: trying to connected to', wsAddress);
 
-    const provider = new WsProvider(wsAddress);
+    const provider = new WsProvider(wsAddress, false);
     const apiPromise = new ApiPromise({provider});
+    apiPromise.connect();
+    dispatch({type: 'CONNECT', payload: apiPromise});
 
     function handleConnect() {
       logger.debug('ChainApiContext: Api connected');
-      setStatus('connected');
+      dispatch({type: 'ON_CONNECT'});
+      queryClient.clear();
```

Comment:
We could also add the network as prefix to the querykeys so we don't lose the cached data when changing the network. But it's out of scope for this issue, we can do it later

Reply:
Sounds good, we also have to make sure not to show very old data. (a refetch should happen while we show the cached data)